File: samples/scoping-samples/simple-stack-example-scoping-java/src/main/java/com/zhuinden/simplestackexamplescoping/features/words/WordController.java
package com.zhuinden.simplestackexamplescoping.features.words;

import com.jakewharton.rxrelay2.BehaviorRelay;
import com.zhuinden.eventemitter.EventEmitter;
import com.zhuinden.eventemitter.EventSource;
import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.Bundleable;
import com.zhuinden.statebundle.StateBundle;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import io.reactivex.Observable;

public class WordController
        implements NewWordFragment.ActionHandler, WordListFragment.ActionHandler, WordListFragment.DataProvider, Bundleable {
    private final BehaviorRelay<List<String>> words = BehaviorRelay.createDefault(Arrays.asList("Bogus",
                                                                                                "Magic",
                                                                                                "Scoping mechanisms"));
    private final EventEmitter<Events> wordEventEmitter = new EventEmitter<>();

    private final Backstack backstack;


    public WordController(Backstack backstack) {
        this.backstack = backstack;
    }

    public EventSource<Events> getEventEmitter() {
        return wordEventEmitter;
    }

    @Nonnull
    @Override
    public StateBundle toBundle() {
        StateBundle stateBundle = new StateBundle();
        //noinspection ConstantConditions
        stateBundle.putStringArrayList("words", new ArrayList<String>(words.getValue()));
        return stateBundle;
    }

    @Override
    public void fromBundle(@Nullable StateBundle bundle) {
        if(bundle != null) {
            words.accept(Collections.unmodifiableList(bundle.getStringArrayList("words")));
        }
    }

    @Override
    public void onAddWordClicked(String word) {
        if(word.isEmpty()) {
            return;
        }
        //noinspection ConstantConditions
        List<String> list = new ArrayList<>(words.getValue());
        list.add(word);
        words.accept(Collections.unmodifiableList(list));
        wordEventEmitter.emit(new Events.NewWordAdded(word));
        backstack.goBack();
    }

    @Override
    public void onAddNewWordClicked() {
        backstack.goTo(NewWordKey.create());
    }

    @Override
    public Observable<List<String>> getWordList() {
        return words;
    }

    public static class Events {
        public static class NewWordAdded
                extends Events {
            private final String word;

            public NewWordAdded(String word) {
                this.word = word;
            }

            public final String getWord() {
                return word;
            }
        }
    }
}

File: samples/scoping-samples/simple-stack-example-scoping-kotlin/src/main/java/com/zhuinden/simplestackexamplescoping/features/words/WordController.kt
package com.zhuinden.simplestackexamplescoping.features.words

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.zhuinden.eventemitter.EventEmitter
import com.zhuinden.eventemitter.EventSource
import com.zhuinden.simplestack.Backstack
import com.zhuinden.simplestack.Bundleable
import com.zhuinden.statebundle.StateBundle

class WordEventEmitter(
    private val eventEmitter: EventSource<WordController.Events>
) : EventSource<WordController.Events> by eventEmitter

class WordController(
    private val backstack: Backstack
) : NewWordFragment.ActionHandler,
    WordListFragment.ActionHandler,
    WordListFragment.DataProvider,
    Bundleable {
    sealed class Events {
        data class NewWordAdded(val word: String) : Events()
    }

    private val wordEventEmitter = EventEmitter<Events>()
    val eventEmitter = WordEventEmitter(wordEventEmitter)

    private val mutableWords: MutableLiveData<List<String>> = MutableLiveData(
        listOf("Bogus", "Magic", "Scoping mechanisms")
    )

    override val wordList: LiveData<List<String>>
        get() = mutableWords

    private fun addWordToList(word: String) {
        mutableWords.run {
            postValue(value!!.toMutableList().also { list -> list.add(word) })
        }
        wordEventEmitter.emit(Events.NewWordAdded(word))
    }

    override fun onAddNewWordClicked() {
        backstack.goTo(NewWordKey)
    }

    override fun onAddWordClicked(word: String) {
        if (word.isNotEmpty()) {
            addWordToList(word)
        }
        backstack.goBack()
    }

    // NOTE: Data is typically in the database, so do this only for non-transient state.
    override fun toBundle(): StateBundle = StateBundle().apply {
        putStringArrayList("words", ArrayList(mutableWords.value))
    }

    override fun fromBundle(bundle: StateBundle?) {
        bundle?.run {
            mutableWords.value = getStringArrayList("words")
        }
    }
}